packages/beacon-node/src/chain/validation/attestation.ts

```diff
@@ -568,41 +571,111 @@ export function verifyHeadBlockAndTargetRoot(
+  chain.metrics?.gossipAttestation.shufflingMiss.inc({caller: regenCaller});
+  try {
+    // for the 1st time of the same epoch and dependent root, it awaits for the regen state
+    // from the 2nd time, it should use the same cached promise and it should reach the above code
+    await chain.regenStateForAttestationVerification(attEpoch, shufflingDependentRoot, attHeadBlock, regenCaller);
   } catch (e) {
     throw new AttestationError(GossipAction.IGNORE, {
       code: AttestationErrorCode.MISSING_STATE_TO_VERIFY_ATTESTATION,
       error: e as Error,
     });
   }
+
+  shuffling = await chain.shufflingCache.get(attEpoch, shufflingDependentRoot);
+  if (shuffling) {
+    chain.metrics?.gossipAttestation.shufflingRegenHit.inc({caller: regenCaller});
+    return shuffling;
+  } else {
+    chain.metrics?.gossipAttestation.shufflingRegenMiss.inc({caller: regenCaller});
+    return null;
+  }
+}
+
+/**
+ * Get dependent root of a shuffling given attestation epoch and head block.
+ */
+export function getShufflingDependentRoot(
+  chain: IBeaconChain,
+  attEpoch: Epoch,
+  blockEpoch: Epoch,
+  attHeadBlock: ProtoBlock
+): RootHex {
+  let shufflingDependentRoot: RootHex;
+  if (blockEpoch === attEpoch) {
+    // current shuffling, this is equivalent to `headState.currentShuffling`
+    // given blockEpoch = attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     ------------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.previous);
+  } else if (blockEpoch === attEpoch - 1) {
+    // next shuffling, this is equivalent to `headState.nextShuffling`
+    // given blockEpoch = n-1, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.current);
+  } else if (blockEpoch < attEpoch - 1) {
+    // this never happens with default chain option of maxSkipSlots = 32, however we still need to handle it
+    // check the verifyHeadBlockAndTargetRoot() function above
+    // given blockEpoch = n-2, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -----------^
+    // shufflingDependentRoot -----^
+    shufflingDependentRoot = attHeadBlock.blockRoot;
+    // use lodestar_gossip_attestation_head_slot_to_attestation_slot metric to track this case
+  } else {
+    // blockEpoch > attEpoch
+    // should not happen, handled in verifyAttestationTargetRoot
+    throw Error(`attestation epoch ${attEpoch} is before head block epoch ${blockEpoch}`);
+  }
+
+  return shufflingDependentRoot;
+}
```

Comment:
Move to beacon-node utils dependentRoot and add sufficient test cases to cover all paths